ballerina-to-openapi/src/main/java/io/ballerina/openapi/service/mapper/interceptor/InterceptorPipeline.java

```diff
@@ -0,0 +1,276 @@
+/*
+ *  Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org).
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.openapi.service.mapper.interceptor;
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.ResourceMethodSymbol;
+import io.ballerina.compiler.api.symbols.ServiceDeclarationSymbol;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.api.symbols.TupleTypeSymbol;
+import io.ballerina.compiler.api.symbols.TypeDefinitionSymbol;
+import io.ballerina.compiler.api.symbols.TypeReferenceTypeSymbol;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.openapi.service.mapper.interceptor.Interceptor.InterceptorType;
+import io.ballerina.openapi.service.mapper.model.ModuleMemberVisitor;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * This {@link InterceptorPipeline} class represents the interceptor pipeline. This class is responsible for building
+ * the interceptor pipeline and extracting the effective return types from the pipeline.
+ *
+ * @since 1.9.0
+ */
+public class InterceptorPipeline {
+
+    private Interceptor initReqInterceptor = null;
+    private Interceptor initResInterceptor = null;
+    private final SemanticModel semanticModel;
+
+    private InterceptorPipeline(SemanticModel semanticModel) {
+        this.semanticModel = semanticModel;
+    }
+
+    public static InterceptorPipeline build(ServiceDeclarationNode serviceDefinition, SemanticModel semanticModel,
+                                            ModuleMemberVisitor moduleMemberVisitor) {
+        List<Interceptor> interceptors = buildInterceptors(serviceDefinition, semanticModel, moduleMemberVisitor);
+
+        InterceptorPipeline pipeline = new InterceptorPipeline(semanticModel);
+        if (!interceptors.isEmpty()) {
+            Interceptor prevInterceptor = interceptors.get(0);
+            if (prevInterceptor.getType().equals(InterceptorType.REQUEST)) {
+                pipeline.initReqInterceptor = prevInterceptor;
+            }
+            if (prevInterceptor instanceof ResponseInterceptor) {
+                pipeline.initResInterceptor = prevInterceptor;
```

Comment:
These 2 scenarios does not happen at once right?

Reply:
Yes, it cannot. Are you suggesting to have a `else if` here?  I thought this is more readable.

Reply:
but i think this gives the impression that a single flow will run both the checks. anyway i'm ok with both options.

Reply:
Fixed via [df3d6fe](https://github.com/ballerina-platform/openapi-tools/pull/1614/commits/df3d6fe5ce05c76ff9a96cc59877f98d9f0cfa29)